### PR TITLE
[prometheus] Add readiness/liveness/startup probes to config-reloader container

### LINKED ISSUE #4623: [prometheus] Liveness and Readiness probes missing
### Is your feature request related to a problem ?

I've a Kyverno rule that checks if all containers have Readiness and Liveness probes defined. It would be nice if "reload" container in Prometheus deployment had these probes enabled as others container.

### Describe the solution you'd like.

Possibility to insert Readiness and Liveness probes.

### Describe alternatives you've considered.

NONE

### Additional context.

_No response_

### PR DESCRIPTION
**Pull request by luigi-bitonti**
#### What this PR does / why we need it
Readiness, liveness and startup probes were added to "configmapReload" container in Prometheus pod.

#### Which issue this PR fixes

*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*

- fixes #4623

#### Checklist

- [X] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [X] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

### CODE REVIEW

**Review by zeritti — CHANGES_REQUESTED** [review]
Thank you, @luigi-bitonti, for your PR. Let me suggest a few changes below.

The first thing is to realise that config-reloader does not open a listening socket. We have to tell it to do so with its flag `--listen-address`. This flag has to be added to the command args and the container's `ports` field introduced. With this, we can use the httpGet probe by default. 

It is ok to reuse the code already present in the chart. We can do more with less, though. Consider these proposed fields in the values file.

```yaml
configmapReload:
  prometheus:
    containerPort: 8080        # port to listen on, argument to the flag mentioned above
    containerPortName: metrics # port name is always useful

    # We can provide access to the whole objects and take them over in templates
    livenessProbe:
      httpGet:
        path: /healthz
        port: metrics
        scheme: HTTP
      periodSeconds: 10
      initialDelaySeconds: 2

    readinessProbe:
      httpGet:
        path: /healthz
        port: metrics
        scheme: HTTP
      periodSeconds: 10

    # If we wish to provide the startup probe with a toggle. 
    startupProbe:
      enabled: false
      httpGet:
        path: /healthz
        port: metrics
        scheme: HTTP
      periodSeconds: 10
```

These fields would have to be implemented in `deploy.yaml` and `sts.yaml`. What do you think?

**Comment by luigi-bitonti** [comment]
I tried to adopt your suggests.

File: charts/prometheus/Chart.yaml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 apiVersion: v2
 name: prometheus
 appVersion: v2.53.0
-version: 25.22.1
+version: 25.22.2
```
Comment (zeritti) [inline charts/prometheus/Chart.yaml:4]: We have to make the change minor.
```suggestion
version: 25.23.0
```

File: charts/prometheus/values.yaml [hunk 2]
```diff
@@ -92,6 +93,34 @@ configmapReload:
     ## Security context to be added to configmap-reload container
     containerSecurityContext: {}
 
+    ## Prometheus reloader readiness and liveness probe initial delay and timeout
```
Comment (zeritti) [inline charts/prometheus/values.yaml:96]:
```suggestion
    ## Settings for Prometheus reloader's readiness, liveness and startup probes
```

File: charts/prometheus/values.yaml [hunk 3]
```diff
@@ -65,7 +65,8 @@ configmapReload:
       digest: ""
       pullPolicy: IfNotPresent
 
-    # containerPort: 9533
+    containerPort: 8080
+    containerPortName: metrics
```
Comment (zeritti) [inline charts/prometheus/values.yaml:69]: A comment will help:
```suggestion
    ## config-reloader's container port and port name for probes and metrics
    containerPort: 8080
    containerPortName: metrics 
```

**Review by zeritti — CHANGES_REQUESTED** [review]
Thank you, @luigi-bitonti, for your changes. Please, see my comments and suggestions below.

File: charts/prometheus/values.yaml [hunk 4]
```diff
@@ -92,6 +93,34 @@ configmapReload:
     ## Security context to be added to configmap-reload container
     containerSecurityContext: {}
 
+    ## Prometheus reloader readiness and liveness probe initial delay and timeout
+    ## Ref: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/
+    ##
+
+    tcpSocketProbeEnabled: false
```
[the remark this answers is not in the document]
Comment (luigi-bitonti) [reply charts/prometheus/values.yaml:100]: Why did you delete this?

File: charts/prometheus/templates/deploy.yaml [hunk 5]
```diff
@@ -91,7 +94,22 @@ spec:
...
+            {{- toYaml . | nindent 12 }}
           {{- end }}
+          {{- with .Values.configmapReload.prometheus.readinessProbe }}
+          readinessProbe:
+            {{- toYaml . | nindent 12 }}
+          {{- end }}
+          {{- if .Values.configmapReload.prometheus.startupProbe.enabled }}
+          {{- $startupProbe := omit .Values.configmapReload.prometheus.startupProbe "enabled" }}
+          startupProbe:
+            {{- toYaml $startupProbe | nindent 12 }}
```
Comment (zeritti) [inline charts/prometheus/templates/deploy.yaml:112]: Closing the conditional has somehow slipped away.
```suggestion
            {{- toYaml $startupProbe | nindent 12 }}
          {{- end }}
```

File: charts/prometheus/templates/sts.yaml [hunk 6]
```diff
@@ -96,7 +99,22 @@ spec:
...
+            {{- toYaml . | nindent 12 }}
           {{- end }}
+          {{- with .Values.configmapReload.prometheus.readinessProbe }}
+          readinessProbe:
+            {{- toYaml . | nindent 12 }}
+          {{- end }}
+          {{- if .Values.configmapReload.prometheus.startupProbe }}
+          {{- $startupProbe := omit .Values.configmapReload.prometheus.startupProbe "enabled" }}
+          startupProbe:
+            {{- toYaml $startupProbe | nindent 12 }}
```
Comment (zeritti) [inline charts/prometheus/templates/sts.yaml:117]: Closing the conditional has somehow slipped away.
```suggestion
            {{- toYaml $startupProbe | nindent 12 }}
          {{- end }}
```

File: charts/prometheus/values.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (zeritti) [reply charts/prometheus/values.yaml:101]: We do not need this field as the user can set any probe, including tcpSocket (setting default httpGet).

**Review by zeritti — CHANGES_REQUESTED** [review]
Thank you, @luigi-bitonti. Please, see my comments below.

File: charts/prometheus/values.yaml [hunk 7]
```diff
@@ -65,7 +65,9 @@ configmapReload:
       digest: ""
       pullPolicy: IfNotPresent
 
-    # containerPort: 9533
+    ## config-reloader's container port and port name for probes and metrics
+    containerPort: 8080
+    containerPortName: metrics
```
Comment (zeritti) [inline charts/prometheus/values.yaml:70]: Please, remove the trailing space:
```suggestion
    containerPortName: metrics
```

**Review by zeritti — CHANGES_REQUESTED** [review]

**Comment by zeritti** [comment]
Thank you, @luigi-bitonti, for your changes. If I may ask, startupProbe is not enabled by default but should be checked by CI. Will you please enable the startupProbe in the values files 02 and 03 in the `ci` directory?

**Review by zeritti — APPROVED** [review]
Many thanks, @luigi-bitonti, LGTM.

**Pull request merged**